kivy/eventmanager/__init__.py

```diff
@@ -0,0 +1,168 @@
+:meth:`EventManagerBase.dispatch` method. It's up to manager to decide how to
+dispatch them, either going through :attr:`EventManagerBase.window.children`
+and dispatch `on_motion` event to them or by using some different logic. It's
+also up to manager to dispatch grabbed events if widgets grab/ungrab to events
+(see :meth:`~kivy.input.motionevent.MotionEvent.grab` and
+:meth:`~kivy.input.motionevent.MotionEvent.ungrab` methods).
+
+Manager can suggest how event should be dispatched further down the widget tree
+by using different flags. Flags are assigned in
+:attr:`~kivy.input.motionevent.MotionEvent.flags` attribute and inspected in
+:meth:`~kivy.uix.widget.Widget.on_motion` method by using bitwise operations.
+
+Currently there are three dispatch behaviors recognized by the widget:
+
+1. Default dispatch (no flag required) - event will go through widget's
+    `children` list, starting with fist widget in the `children` list until
+    child widget accept the event or last widget registered for that event is
+    reached.
+2. Filtered dispatch (requires :attr:`FILTERED_DISPATCH`) - event will go only
+    through registered child widgets.
+3. No children dispatch (requires :attr:`DONT_DISPATCH`) - event will not be
+    dispatched to child widgets.
+
+Note that window does not have `motion_filter` property and therefore does not
+have a list of filtered widgets from its `children` list.
+'''
+
+DONT_DISPATCH = 1
+'''Flag which suggest that event should not be dispatch to child widgets, but
+only go through method resolution order of
+:meth:`~kivy.uix.widget.Widget.on_motion` so that all super classes can handle
+the event.
+'''
+
+FILTERED_DISPATCH = 2
+'''Flag which suggest that event should be dispatched only to child widgets
+which were previously registered to receive events of the same
+:attr:`~kivy.input.motionevent.MotionEvent.type_id` and not to all
+child widgets.
+'''
+
+
+class EventManagerBase(object):
+    '''Abstract class with methods :meth:`start`, :meth:`stop` and
+    :meth:`dispatch` for specific class to implement.
+
+    Example of the manager receiving touch and hover events::
+
+        class TouchHoverManager(EventManagerBase):
+
+            type_ids = ('touch', 'hover')
+
+            def start(self):
+                # Create additional resources, bind callbacks to self.window
+
+            def dispatch(self, etype, me):
+                if me.type_id == 'touch':
+                    # Handle touch event
+                elif me.type_id == 'hover'
+                    # Handle hover event
+
+            def stop(self):
+                # Release resources
+
+    '''
+
+    type_ids = None
+    '''Override this attribute to declare type ids of events which manager will
+    receive. This attribute will be used by
+    :class:`~kivy.core.window.WindowBase` to know which events to pass to
+    method :meth:`dispatch`.
+
+    .. versionadded:: 2.1.0
+    '''
+
+    window = None
```

Comment:
To avoid using global variable(s) and its global or local import, so that local import in `start` method
```
Window = None

def start(self):
    global Window
    if not Window:
        from kivy.core.window import Window
```
can be replaced with a single line `self.window` and then used.

Seems better for (some) unit tests, as it needs a dummy window and widget objects with `on_motion` event and method so test can be run without window creation.

Reply:
ok, indeed it would better if we remove the side effects of importing kivy modules over time. Alternatively it could just be a kivy property, allowing to use EventLoop.window anywhere needed.
```py
@property 
def window(self):
      from kivy.core.window import Window
      return Window
```

Reply:
So we keep the `window` attribute?

`EventLoop.window` already exists as an attribute and it's set with `EventLoop.set_window` method.

Reply:
Oh, right, sorry, I should have checked.
I think we can keep it, yes.